### ISSUE TITLE
Full port to PHP?

### ISSUE THREAD

**Issue by robiningelbrecht** [tdecker91/puzzle-gen#1]
Any chance we can port this package to PHP? I know there's https://github.com/Cride5/visualcube already, but it kinda seems abandoned AND this library has way more features.

I'd be willing to kickstart this...

**Comment by tdecker91** [comment]
@robiningelbrecht I don't have any plans to build out a php port. You're more than welcome to give it a go. The good news is that there are no third party dependencies used so porting _should_ be trivial.

If you're looking to run this serverside instead of in the web client my intention was to make this runnable in node.js It's possible now by providing a node canvas renderer to the visualizer class as show [here](https://tdecker91.github.io/puzzle-gen/#/advanced?id=custom-renderers). Using that it's possible to build a server application that outputs images much like visualcube did.

**Comment by robiningelbrecht** [comment]
@tdecker91 That server side rendering is indeed what I was looking for, thanks for the tip! Is there a full working example of this in the repo?

**Comment by tdecker91** [comment]
nothing in the repo so far. 
The only thing I wrote was that example in the `custom renderers` section of the docs to give an idea of what was possible.

There is an implementation for the `drawArrow` function in the html canvas renderer [here](https://github.com/tdecker91/puzzle-gen/blob/master/src/rendering/htmlCanvasRenderer.ts#L84-L108)

With those that'd be what my first implementation of the "node canvas renderer" would be.

I intentionally left it out since it added a node dependency for [canvas](https://www.npmjs.com/package/canvas) and I was at first focused on working in the client. Plus I didn't want to think about managing builds of the library for node/web 😅. In that case it would be good to separate the core puzzle rendering away from the web and/or node library.

**Comment by robiningelbrecht** [comment]
FYI: I started porting this library to PHP. I finished the basic cube functionality, but still need to add all other puzzle types, masks and arrows

First of all, thanks for the great work you did, the maths, clean code. Great job!

Second, I hope you don't mind me porting this to PHP? I'll definitly reference this GitHub repo ofcourse.

Third, If you want to take a look at the code: https://github.com/robiningelbrecht/puzzle-generator. Still need to add docs though. But the idea is to be able to generate cubes by navigating to `http://localhost:9090/cube?cube[algorithm]=B U R'&cube[size]=4&size=300` for example
